### PR TITLE
records: do not allow stale records updates over HTTP

### PR DESCRIPTION
**Pull request by harunurhan**
* Installs a fork of `invenio-records-rest` which does weak etag checks
* Modifies record-editor to send `If-Match`
* Uses `vary` header instead of `Content-Type` aware `Etag`
which fixes `Etag` and `If-None-Match` based HTTP caching

Ref #1387

### CODE REVIEW

File: backend/inspirehep/editor/views.py [hunk 1]
```diff
@@ -50,9 +50,12 @@ def revert_to_revision(endpoint, pid_value):
 def get_record_and_schema(endpoint, pid_value):
     pid_type = PidStoreBase.get_pid_type_from_endpoint(endpoint)
     record = InspireRecord.get_record_by_pid_value(pid_value, pid_type)
-    return jsonify(
-        {"record": {"metadata": record}, "schema": load_schema(record["$schema"])}
-    )
+    json = {"record": {"metadata": record}, "schema": load_schema(record["$schema"])}
+
+    response = make_response(json)
+    set_headers_for_record_caching_and_concurrency(response, record)
```
Comment (harunurhan) [inline backend/inspirehep/editor/views.py:56]: this could potentially cause a state cache, when schema is changed but not the record.

File: backend/inspirehep/utils.py [hunk 2]
```diff
@@ -45,3 +45,8 @@ def hash_data(data):
     if data:
         return hashlib.md5(data).hexdigest()
     raise ValueError("Data for hashing cannot be empty")
+
+
+def set_headers_for_record_caching_and_concurrency(response, record):
+    response.set_etag(str(record.revision_id), weak=True)
+    response.vary = "Content-Type,Accept,Accept-Encoding,Accept-Language"
```
Comment (michamos) [inline backend/inspirehep/utils.py:52]: I don't think you need the `content-type`, it seems vary is used to add additional *request* headers to the cache key.

File: backend/tests/integration/records/views/test_views.py [hunk 3]
```diff
@@ -67,3 +67,36 @@ def test_returns_deleted_pid_error_if_not_cataloger(inspire_app):
...
+        stale_put_response = client.put(
+            f"/conferences/{record_control_number}",
+            content_type="application/json",
+            data=json.dumps(
+                faker.record("con", data={"control_number": record_control_number})
+            ),
+            headers={"If-Match": etag},
+        )
+
+    assert stale_put_response.status_code == 412
```
Comment (michamos) [inline backend/tests/integration/records/views/test_views.py:102]: maybe also add a test for a 304 on a `GET` with matching `If-None-Modified`?

**Review by drjova — APPROVED** [review]
👏
